commit 2add05ddf6c15e71369970868a07cfc394e677e9 (refs/changes/73/2573/1)
Author: Sam Gwydir <sam.gwydir@joyent.com>
Date:   2017-09-14T18:12:05+00:00 (2 years, 1 month ago)
    
    IMGAPI-640 Want sdc:operator-script to run before smf services start during image creation.

diff --git a/overlay/generic/lib/svc/manifest/system/mdata-operator-script.xml b/overlay/generic/lib/svc/manifest/system/mdata-operator-script.xml
new file mode 100644
index 00000000..7062761c
--- /dev/null
+++ b/overlay/generic/lib/svc/manifest/system/mdata-operator-script.xml
@@ -0,0 +1,29 @@
+<?xml version="1.0"?>
+<!DOCTYPE service_bundle SYSTEM "/usr/share/lib/xml/dtd/service_bundle.dtd.1">
+<service_bundle type="manifest" name="mdata-operator-script">
+  <service name="smartdc/mdata-operator-script" type="service" version="1">
+    <dependency name="filesystem" grouping="require_all" restart_on="error" type="service">
+      <service_fmri value="svc:/system/filesystem/minimal" />
+    </dependency>
+    <dependent name='single-user' grouping='require_all' restart_on='none'>
+      <service_fmri value='svc:/milestone/single-user:default' />
+    </dependent>
+    <property_group name="startd" type="framework">
+      <propval name="duration" type="astring" value="transient" />
+      <propval name="ignore_error" type="astring" value="core,signal" />
+    </property_group>
+    <instance name="operator-script" enabled="true">
+      <dependency name="loopback" grouping="require_all" restart_on="error" type="service">
+        <service_fmri value="svc:/network/loopback:default" />
+      </dependency>
+      <exec_method type="method" name="start" exec="/lib/svc/method/mdata-operator-script" timeout_seconds="0" />
+      <exec_method type="method" name="stop" exec=":true" timeout_seconds="60" />
+    </instance>
+    <stability value="Evolving" />
+    <template>
+      <common_name>
+        <loctext xml:lang="C">Joyent SDC operator script handler</loctext>
+      </common_name>
+    </template>
+  </service>
+</service_bundle>
diff --git a/overlay/generic/lib/svc/method/mdata-execute b/overlay/generic/lib/svc/method/mdata-execute
index fca08ffb..dec5d98c 100755
--- a/overlay/generic/lib/svc/method/mdata-execute
+++ b/overlay/generic/lib/svc/method/mdata-execute
@@ -35,15 +35,6 @@ if [ -f /var/svc/provisioning ]; then
     mv /var/svc/provision{ing,_success}
 fi
 
-if [[ -x /var/svc/mdata-operator-script ]]; then
-    /var/svc/mdata-operator-script
-    operator_script_exit=$?
-    if [[ ${operator_script_exit} -gt 0 ]]; then
-        echo "WARNING: operator-script failed: exited ${operator_script_exit}" \
-            >&2
-    fi
-fi
-
 user_script_exit=${SMF_EXIT_OK}
 if [ -x /var/svc/mdata-user-script ]; then
     /var/svc/mdata-user-script
diff --git a/overlay/generic/lib/svc/method/mdata-operator-script b/overlay/generic/lib/svc/method/mdata-operator-script
new file mode 100755
index 00000000..21ead778
--- /dev/null
+++ b/overlay/generic/lib/svc/method/mdata-operator-script
@@ -0,0 +1,74 @@
+#!/usr/bin/bash
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License, Version 1.0 only
+# (the "License").  You may not use this file except in compliance
+# with the License.
+#
+# You can obtain a copy of the license at http://smartos.org/CDDL
+#
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file.
+#
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+# Copyright (c) 2015, Joyent, Inc. All rights reserved.
+#
+
+export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+export PATH=/usr/bin:/usr/sbin:$PATH
+
+set -o xtrace
+
+. /lib/svc/share/smf_include.sh
+smf_is_globalzone && exit ${SMF_EXIT_OK}
+
+if [ ! -x /usr/sbin/mdata-get ]; then
+  echo "Metadata mdata-get tool not found."
+  exit ${SMF_EXIT_ERR_FATAL}
+fi
+
+echo "Retrieving metadata operator-script..."
+/usr/sbin/mdata-get sdc:operator-script >/var/svc/mdata-operator-script.new
+case $? in
+  0)
+    echo "Metadata operator-script successfuly retrieved."
+    mv /var/svc/mdata-operator-script{.new,}
+    chmod +x /var/svc/mdata-operator-script
+    ;;
+  1)
+    echo "Metadata operator-script not defined."
+    rm -f /var/svc/mdata-operator-script{,.new}
+    ;;
+  *)
+    echo "Metadata couldn't be retrieved."
+    exit ${SMF_EXIT_ERR_FATAL}
+    ;;
+esac
+
+
+# If we got as far as running the user-script the 'provision' was a success
+# from here out a failure will leave the zone running.
+if [ -f /var/svc/provisioning ]; then
+    mv /var/svc/provision{ing,_success}
+fi
+
+if [[ -x /var/svc/mdata-operator-script ]]; then
+    /var/svc/mdata-operator-script
+    operator_script_exit=$?
+    if [[ ${operator_script_exit} -gt 0 ]]; then
+        echo "WARNING: operator-script failed: exited ${operator_script_exit}" \
+            >&2
+    fi
+fi
+
+exit ${user_script_exit}
diff --git a/overlay/generic/manifest b/overlay/generic/manifest
index 2b621fe5..cd1eb433 100644
--- a/overlay/generic/manifest
+++ b/overlay/generic/manifest
@@ -77,6 +77,7 @@ d lib/svc/manifest/system/filesystem 0755 root sys
 f lib/svc/manifest/system/filesystem/joyent-fs.xml 0444 root sys
 f lib/svc/manifest/system/filesystem/minimal-fs.xml 0444 root sys
 f lib/svc/manifest/system/mdata.xml 0444 root sys
+f lib/svc/manifest/system/mdata-operator-script.xml 0444 root sys
 f lib/svc/manifest/system/smartdc-config.xml 0444 root sys
 f lib/svc/manifest/system/smartdc-init.xml 0444 root sys
 f lib/svc/manifest/system/smartdc-ur.xml 0444 root sys
@@ -88,6 +89,7 @@ f lib/svc/method/fs-usr 0555 root bin
 f lib/svc/method/identity-node 0555 root bin
 f lib/svc/method/ipfilter 0555 root sys
 f lib/svc/method/manifest-import 0555 root bin
+f lib/svc/method/mdata-operator-script 0555 root bin
 f lib/svc/method/mdata-execute 0555 root bin
 f lib/svc/method/mdata-fetch 0555 root bin
 f lib/svc/method/net-physical 0555 root sys
diff --git a/overlay/generic/usr/lib/brand/joyent-minimal/manifests b/overlay/generic/usr/lib/brand/joyent-minimal/manifests
index 955d4670..544b60b6 100644
--- a/overlay/generic/usr/lib/brand/joyent-minimal/manifests
+++ b/overlay/generic/usr/lib/brand/joyent-minimal/manifests
@@ -125,3 +125,4 @@ milestone/name-services.xml	enabled
 system/early-manifest-import.xml	disabled
 system/manifest-import.xml	disabled
 system/mdata.xml	enabled
+system/mdata-operator-script.xml	enabled
